### PR TITLE
Masterbar: updates the default URL for Plans

### PR DESCRIPTION
**Pull request by keoshi**
#### Changes proposed in this Pull Request:

* Updates the default link in `Upgrades → Plans` to "Plans" instead of "My Plan".
* Updates tests accordingly.
* Related Calypso changes: https://github.com/Automattic/wp-calypso/pull/53884

#### Jetpack product discussion

p7DVsv-ago-p2#comment-36411

#### Does this pull request change what data or activity we track or use?

No.

#### Testing instructions:

* You can either:
  * Apply D63148-code in your sandbox to test this site in WordPress.com.
  * Switch to the branch of this PR in your local environment and test an Atomic site on WordPress.com.
* Ensure you're redirected to “Plans” when navigating into `Upgrades → Plans`.

### CODE REVIEW

**Review by cpapazoglou — COMMENTED** [review]
Thanks @keoshi  for working on this!
LGTM! You will need to update the rest of the tests too though:

- https://github.com/Automattic/jetpack/blob/e694c88492f3bf7113a2b6b2b2b43875decd1852/projects/plugins/jetpack/tests/php/modules/masterbar/test-class-atomic-admin-menu.php#L303
- https://github.com/Automattic/jetpack/blob/2831b9bd61307a81ef5ca50f95c5bba7c8df8d1e/projects/plugins/jetpack/tests/php/modules/masterbar/test-class-wpcom-admin-menu.php#L254

**Comment by leogermani** [comment]
We should probably use Jetpack Redirects (PCYsg-pY7-p2) in this link.

File: projects/plugins/jetpack/modules/masterbar/admin-menu/class-admin-menu.php [hunk 1]
```diff
@@ -188,7 +188,7 @@ public function add_upgrades_menu( $plan = null ) {
 			add_menu_page( __( 'Upgrades', 'jetpack' ), $site_upgrades, 'manage_options', 'paid-upgrades.php', null, 'dashicons-cart', 4 );
 		}
 
-		add_submenu_page( 'paid-upgrades.php', __( 'Plans', 'jetpack' ), __( 'Plans', 'jetpack' ), 'manage_options', 'https://wordpress.com/plans/my-plan/' . $this->domain, null, 1 );
+		add_submenu_page( 'paid-upgrades.php', __( 'Plans', 'jetpack' ), __( 'Plans', 'jetpack' ), 'manage_options', 'https://wordpress.com/plans/' . $this->domain, null, 1 );
```
Comment (keoshi) [inline projects/plugins/jetpack/modules/masterbar/admin-menu/class-admin-menu.php:191]: @leogermani I'm afraid this is a bit over my head. Would this be the way to do it?

```suggestion
		add_submenu_page( 'paid-upgrades.php', __( 'Plans', 'jetpack' ), __( 'Plans', 'jetpack' ), 'manage_options', esc_url( Redirect::get_url( 'calypso-backups' ) . $this->domain ), null, 1 );
```

Comment (leogermani) [reply projects/plugins/jetpack/modules/masterbar/admin-menu/class-admin-menu.php:191]: Replied in p1624380899326900-slack-CBG1CP4EN

Comment (keoshi) [reply projects/plugins/jetpack/modules/masterbar/admin-menu/class-admin-menu.php]: Updated in https://github.com/Automattic/jetpack/pull/20141/commits/b5943cb49d07c1973f3e1c3315d61b11f8f90af0 — thanks for all the guidance, @leogermani !

**Comment by sdixon194** [comment]
Hmm, I think something is not right on my end. The upgrades and plans links didn't work for me at all with this applied:

Screen Capture on 2021-06-23 at 11-03-50

I'm not sure if maybe I'm testing incorrectly?

**Comment by keoshi** [comment]
@sdixon194 Could it be that it also needs the Calypso PR to work fully? https://github.com/Automattic/wp-calypso/pull/53884

**Comment by keoshi** [comment]
@sdixon194 @leogermani Can I ask you both for a quick review on this and https://github.com/Automattic/wp-calypso/pull/53884 so we can get this changed shipped this week, please?

**Comment by sdixon194** [comment]
@keoshi I tried again this time while checked out on the Calypso branch. I still can't get the PR to work :/ 

To recap my steps to make sure I'm testing correctly:

- checked out calypso branch
- yarn and yarn start
- opened http://calypso.localhost:3000/ and selected my atomic test site with the beta plugin and test branch active
- Plan page link doesn’t work.

Hovering over the Plan URL, it looks like it's registering the same link as whatever my currently active page is:

Screen Shot on 2021-07-26 at 12:12:43

**Comment by keoshi** [comment]
@cpapazoglou As the expert on Nav Unification stuff, would you mind giving us a hand with this one and its instructions?

**Comment by cpapazoglou** [comment]
@sdixon194 
I am not sure which calypso branch you have checked out, this is a Jetpack PR. You can either apply D63148-code in your sandbox and test this site in WordPress.com or Jetpack Beta with `update/admin-menu-plans-default-url` and test that Atomic site in Wordpress.com. I am not sure why `update/admin-menu-plans-default-url` is currently not showing up in Jetpack Beta, checking out this branch in my local Atomic jetpack worked though.

**Atomic**

WPAdmin | Calypso
-------|------
 | 

**Simple**
WPAdmin | Calypso
-------|------
|

In all cases I landed in `/plans/[DOMAIN]` which what is expected @keoshi right?

**Comment by keoshi** [comment]
Exactly that, @cpapazoglou ! Thanks so much for jumping in with the clarification. I'll make sure to update the instructions.

**Comment by anomiex** [comment]
> I am not sure why `update/admin-menu-plans-default-url` is currently not showing up in Jetpack Beta

I see it there, and I see no sign that has changed in the 4 hours since you posted. Possibly it's the issue fixed by #20417 (the search not handling `-` correctly), which will be included in 3.0.2.

I was hoping to get review on #20432 before doing another release, but review or not I'll try to get a Beta Plugin release out sometime this week.

**Comment by samiff** [comment]
Having a bit of trouble fully testing this one 🤔  

* Applied D63148-code to my simple sandbox site.
* The wp-admin side seems to work well:



* That resolves to `https://wordpress.com/plans/__URL__.wordpress.com?site=__URL__.wordpress.com` for me (the Plans page).
* However, next I select say Posts so I'm on `https://wordpress.com/posts/__URL__.wordpress.com`, and hovering over the Upgrades --> Plans now doesn't seem to work, just copies whatever page I'm viewing in Calypso (clicking doesn't navigate anywhere):



Have I misunderstood what is being tested here or testing setup? CC: @keoshi @cpapazoglou

**Comment by cpapazoglou** [comment]
> Having a bit of trouble fully testing this one 🤔
> 
> * Applied D63148-code to my simple sandbox site.
> * The wp-admin side seems to work well:
> 
> 
> 
> * That resolves to `https://wordpress.com/plans/__URL__.wordpress.com?site=__URL__.wordpress.com` for me (the Plans page).
> * However, next I select say Posts so I'm on `https://wordpress.com/posts/__URL__.wordpress.com`, and hovering over the Upgrades --> Plans now doesn't seem to work, just copies whatever page I'm viewing in Calypso (clicking doesn't navigate anywhere):
> 
> 
> 
> Have I misunderstood what is being tested here or testing setup? CC: @keoshi @cpapazoglou

I suppose you are sandboxing the WPCOM api too, right? Can you please share the `admin-menu` api response?

**Review by cpapazoglou — DISMISSED** [review]
I am not sure why I couldn't reproduce the problem at first but it's apparent now (I was probably seeing cached results).

Calypso accepts only WP Admin external urls

https://github.com/Automattic/wp-calypso/blob/da1a68dc58bd6d251dbdd5246bc5944865e967f0/client/state/data-layer/wpcom/sites/admin-menu/index.js#L22-L43

`https://jetpack.com/redirect/?source=masterbar-plans&site=cpapazoglousandbox.wordpress.com` doesn't get past the validation and results to `''`

Why do we use `Redirect::get_url( 'masterbar-plans' )` in the first place? If it's totally needed, we should firstly change Calypso to also accept `jetpack.com` urls.

**Comment by keoshi** [comment]
@leogermani Do you think we really need the JP.com redirection here? What do we get for free in that scenario, tracking?

I'd personally love to see this merged along with https://github.com/Automattic/wp-calypso/pull/53884 as quickly as possible in order to track if we're losing any CR because of the revert ([more info here](p7DVsv-ago-p2#comment-36411)).

**Comment by kraftbj** [comment]
@keoshi We get some tracking, yes. Notable too is we get the ability to update the URL without requiring a client-side update, able to update the URL between Calypso blue and green dynamically, etc. If we ever need to update the URL in the future, the redirect allows us to support any version of Jetpack (that has the redirect) vs needing to have a bificated experience with older versions going to an old URL, etc.

Overall, I'd say Calypso should accept Jetpack.com URLs or do something along the lines of opening up a WordPress.com/redirect URL that is processed the same as the Jetpack.com ones, etc. The latter would be a bigger thing though.

**Comment by keoshi** [comment]
> the ability to update the URL without requiring a client-side update

Ha, that's right — I forgot about that one. Can we add support for `jetpack.com` URLs then?

Looping in @Automattic/jetpack-voyager because this is proving to be more complex than initially planned.

**Comment by cpapazoglou** [comment]
> > the ability to update the URL without requiring a client-side update
> 
> Ha, that's right — I forgot about that one. Can we add support for `jetpack.com` URLs then?
> 
> Looping in @Automattic/jetpack-voyager because this is proving to be more complex than initially planned.

In that case, I don't see any problems in adding support for `jetpack.com` URLs here https://github.com/Automattic/wp-calypso/blob/da1a68dc58bd6d251dbdd5246bc5944865e967f0/client/state/data-layer/wpcom/sites/admin-menu/index.js#L22-L43

**Comment by mattgawarecki** [comment]
@cpapazoglou I reproduced the same issue you describe, and I agree that what you're suggesting would fix it. Do we need to check with any security or Calypso-centric teams to get their blessing?

**Comment by cpapazoglou** [comment]
> @cpapazoglou I reproduced the same issue you describe, and I agree that what you're suggesting would fix it. Do we need to check with any security or Calypso-centric teams to get their blessing?

Yeah, a calypsop2 post would be great!

**Comment by rcanepa** [comment]
@keoshi I wrote a quick proof of concept that makes Calypso accepts Jetpack Redirect URLs: https://github.com/Automattic/wp-calypso/pull/55138. If you already have an open PR, feel free to cherry-pick the commit and close mine.

Finally, I didn't test all the cases, so I'll leave that to you.

**Comment by samiff** [comment]
I've updated the PR @rcanepa started: https://github.com/Automattic/wp-calypso/pull/55138/

If someone could review that, it'd be much appreciated.

**Review by cpapazoglou — DISMISSED** [review]
Tested this with https://github.com/Automattic/wp-calypso/pull/55138 and while it works, when I land to the plans page the correct menus are **not highlighted / active** to the left



This is due to Calypso not being able to extract the correct path from the Jetpack link `https://jetpack.com/redirect/?source=masterbar-plans&site=cpapfree.wordpress.com`. I am still not sure why we should use that kind of links in Calypso simple sites. Is there a chance we need that kind of redirect only for Jetpack sites?

cc: @rcanepa

**Comment by rcanepa** [comment]
> This is due to Calypso not being able to extract the correct path from the Jetpack link `https://jetpack.com/redirect/?source=masterbar-plans&site=cpapfree.wordpress.com`. I am still not sure why we should use that kind of links in Calypso simple sites. Is there a chance we need that kind of redirect only for Jetpack sites?

I honestly don't know why. @keoshi, I see that you added an entry for that source in Jetpack.com Redirect Manager. Do you happen to know why we are using a redirect rather than a direct link?

I understand that we prefer redirects over direct links when we deal with links inside the plugin so we can change them without requiring a new release. So, I don't see why we need them in Calypso.

**Comment by keoshi** [comment]
@rcanepa Yeah, I can't find a reason why not to use simple links here. Pinging @leogermani since he'll know a lot more about this.

**Review by samiff — APPROVED** [review]

**Comment by samiff** [comment]
@keoshi This PR should be good to land now 👍   Let me know if you need a hand with anything.

**Comment by leogermani** [comment]
r233484-wpcom

**Pull request merged**